# Exemplos de Regressão em Ruby

### Notebook de Rubens Minoru Andako Bueno
#### [VAGAS.com](http://www.vagas.com.br/)

## Sumário

1. [Bibliotecas](#Bibliotecas)

2. [Desafio](#Desafio)

3. [Questões](#Questões)

4. [Passo 1: Visualizando os dados](#Passo-1:-Visualizando-os-dados)

5. [Passo 2: Correlação](#Passo-2:-Correlação)

6. [Passo 3: Regressão](#Passo-3:-Regressão)

7. [Passo 4: Previsão](#Passo-4:-Previsão)

## Bibliotecas

[[ Volte ao topo ]](#Sumário)

Os principais pacotes que nós iremos usar são:

* **daru**: Biblioteca para armazenamento, analise, manipulação e visualização de dados em Ruby.
* **nyaplot**: Biblioteca para gerar gráficos interativos em Ruby.
* **linear-regression**: Biblioteca que implementa algoritmos para modelagem de regressão linear.

## Desafio

[[ Volte ao topo ]](#Sumário)

Quais são os **atributos**?
- **TV**: Valor de divulgação (em milhares de dólares) gasto na TV para divulgação de um dado produto em um dado mercado
- **Radio**: Valor de divulgação gasto na Rádio
- **Jornal**: Valor de divulgação gasto no Jornal

Qual a **resposta**?
- **Sales**: Vendas de um dado produto (em milhares) em um dado mercado

Vamos dar uma olhada nos dados:

In [55]:
require 'daru'

advertising_data = Daru::DataFrame.from_csv('advertising-data.csv')
advertising_data.head(5)

#<Daru::DataFrame(5x5)>
                            TV     Radio Newspaper     Sales
         0         1     230.1      37.8      69.2      22.1
         1         2      44.5      39.3      45.1      10.4
         2         3      17.2      45.9      69.3       9.3
         3         4     151.5      41.3      58.5      18.5
         4         5     180.8      10.8      58.4      12.9

São 200 **observações**, portanto, 200 mercados no dataset.

## Questões

[[ Volte ao topo ]](#Sumário)

Vamos fingir que você trabalhe para uma empresa que manufatura e vende este produto. A empresa pode questionar o seguinte: Baseado nesses dados, como eles deveriam gastar no futuro?

Esta questão pode levantar questões mais específicas, como:
1. Existe relação entre propaganda e venda?
2. Quão forte é a relação?
3. Que tipo de propaganda contribui para vendas?
4. Qual o efeito de cada tipo de propaganda nas vendas?
5. Dado um gasto em propaganda para um mercado, podemos prever as vendas?

Vamos explorar essas questões!

## Passo 1: Visualizando os dados

[[ Volte ao topo ]](#Sumário)

In [56]:
advertising_data.describe

#<Daru::DataFrame(5x5)>
                        Newspaper      Radio      Sales         TV
      count        200        200        200        200        200
       mean      100.5 30.5539999 23.2640000 14.0225000   147.0425
        std 57.8791845 21.7786208 14.8468091 5.21745656 85.8542363
        min          1        0.3          0        1.6        0.7
        max        200        114       49.6         27      296.4

In [57]:
plot = Nyaplot::Plot.new
sc = plot.add_with_df(advertising_data.to_nyaplotdf, :scatter, 'TV', 'Sales')
plot.x_label("TV")
plot.y_label("Sales")
plot.show()

#<Nyaplot::Frame:0x007fe2c9eeaaa8 @properties={:panes=>[#<Nyaplot::Plot:0x007fe2ca085e08 @properties={:diagrams=>[#<Nyaplot::Diagram:0x007fe2c9eeb250 @properties={:type=>:scatter, :options=>{:x=>"TV", :y=>"Sales"}, :data=>"a2f56b13-00ca-44d5-93d3-40fcc6bf45c9"}, @xrange=[0.7, 296.4], @yrange=[1.6, 27]>], :options=>{:x_label=>"TV", :y_label=>"Sales", :zoom=>true, :width=>700, :xrange=>[0.7, 296.4], :yrange=>[1.6, 27]}}>], :data=>{"a2f56b13-00ca-44d5-93d3-40fcc6bf45c9"=>#<Nyaplot::DataFrame:0x007fe2c9f1a460 @name="a2f56b13-00ca-44d5-93d3-40fcc6bf45c9", @rows=[{:""=>1, :TV=>230.1, :Radio=>37.8, :Newspaper=>69.2, :Sales=>22.1}, {:""=>2, :TV=>44.5, :Radio=>39.3, :Newspaper=>45.1, :Sales=>10.4}, {:""=>3, :TV=>17.2, :Radio=>45.9, :Newspaper=>69.3, :Sales=>9.3}, {:""=>4, :TV=>151.5, :Radio=>41.3, :Newspaper=>58.5, :Sales=>18.5}, {:""=>5, :TV=>180.8, :Radio=>10.8, :Newspaper=>58.4, :Sales=>12.9}, {:""=>6, :TV=>8.7, :Radio=>48.9, :Newspaper=>75, :Sales=>7.2}, {:""=>7, :TV=>57.5, :Radio=>32.8, :Newspaper=>23.5, :Sales=>11.8}, {:""=>8, :TV=>120.2, :Radio=>19.6, :Newspaper=>11.6, :Sales=>13.2}, {:""=>9, :TV=>8.6, :Radio=>2.1, :Newspaper=>1, :Sales=>4.8}, {:""=>10, :TV=>199.8, :Radio=>2.6, :Newspaper=>21.2, :Sales=>10.6}, {:""=>11, :TV=>66.1, :Radio=>5.8, :Newspaper=>24.2, :Sales=>8.6}, {:""=>12, :TV=>214.7, :Radio=>24, :Newspaper=>4, :Sales=>17.4}, {:""=>13, :TV=>23.8, :Radio=>35.1, :Newspaper=>65.9, :Sales=>9.2}, {:""=>14, :TV=>97.5, :Radio=>7.6, :Newspaper=>7.2, :Sales=>9.7}, {:""=>15, :TV=>204.1, :Radio=>32.9, :Newspaper=>46, :Sales=>19}, {:""=>16, :TV=>195.4, :Radio=>47.7, :Newspaper=>52.9, :Sales=>22.4}, {:""=>17, :TV=>67.8, :Radio=>36.6, :Newspaper=>114, :Sales=>12.5}, {:""=>18, :TV=>281.4, :Radio=>39.6, :Newspaper=>55.8, :Sales=>24.4}, {:""=>19, :TV=>69.2, :Radio=>20.5, :Newspaper=>18.3, :Sales=>11.3}, {:""=>20, :TV=>147.3, :Radio=>23.9, :Newspaper=>19.1, :Sales=>14.6}, {:""=>21, :TV=>218.4, :Radio=>27.7, :Newspaper=>53.4, :Sales=>18}, {:""=>22, :TV=>237.4, :Radio=>5.1, :Newspaper=>23.5, :Sales=>12.5}, {:""=>23, :TV=>13.2, :Radio=>15.9, :Newspaper=>49.6, :Sales=>5.6}, {:""=>24, :TV=>228.3, :Radio=>16.9, :Newspaper=>26.2, :Sales=>15.5}, {:""=>25, :TV=>62.3, :Radio=>12.6, :Newspaper=>18.3, :Sales=>9.7}, {:""=>26, :TV=>262.9, :Radio=>3.5, :Newspaper=>19.5, :Sales=>12}, {:""=>27, :TV=>142.9, :Radio=>29.3, :Newspaper=>12.6, :Sales=>15}, {:""=>28, :TV=>240.1, :Radio=>16.7, :Newspaper=>22.9, :Sales=>15.9}, {:""=>29, :TV=>248.8, :Radio=>27.1, :Newspaper=>22.9, :Sales=>18.9}, {:""=>30, :TV=>70.6, :Radio=>16, :Newspaper=>40.8, :Sales=>10.5}, {:""=>31, :TV=>292.9, :Radio=>28.3, :Newspaper=>43.2, :Sales=>21.4}, {:""=>32, :TV=>112.9, :Radio=>17.4, :Newspaper=>38.6, :Sales=>11.9}, {:""=>33, :TV=>97.2, :Radio=>1.5, :Newspaper=>30, :Sales=>9.6}, {:""=>34, :TV=>265.6, :Radio=>20, :Newspaper=>0.3, :Sales=>17.4}, {:""=>35, :TV=>95.7, :Radio=>1.4, :Newspaper=>7.4, :Sales=>9.5}, {:""=>36, :TV=>290.7, :Radio=>4.1, :Newspaper=>8.5, :Sales=>12.8}, {:""=>37, :TV=>266.9, :Radio=>43.8, :Newspaper=>5, :Sales=>25.4}, {:""=>38, :TV=>74.7, :Radio=>49.4, :Newspaper=>45.7, :Sales=>14.7}, {:""=>39, :TV=>43.1, :Radio=>26.7, :Newspaper=>35.1, :Sales=>10.1}, {:""=>40, :TV=>228, :Radio=>37.7, :Newspaper=>32, :Sales=>21.5}, {:""=>41, :TV=>202.5, :Radio=>22.3, :Newspaper=>31.6, :Sales=>16.6}, {:""=>42, :TV=>177, :Radio=>33.4, :Newspaper=>38.7, :Sales=>17.1}, {:""=>43, :TV=>293.6, :Radio=>27.7, :Newspaper=>1.8, :Sales=>20.7}, {:""=>44, :TV=>206.9, :Radio=>8.4, :Newspaper=>26.4, :Sales=>12.9}, {:""=>45, :TV=>25.1, :Radio=>25.7, :Newspaper=>43.3, :Sales=>8.5}, {:""=>46, :TV=>175.1, :Radio=>22.5, :Newspaper=>31.5, :Sales=>14.9}, {:""=>47, :TV=>89.7, :Radio=>9.9, :Newspaper=>35.7, :Sales=>10.6}, {:""=>48, :TV=>239.9, :Radio=>41.5, :Newspaper=>18.5, :Sales=>23.2}, {:""=>49, :TV=>227.2, :Radio=>15.8, :Newspaper=>49.9, :Sales=>14.8}, {:""=>50, :TV=>66.9, :Radio=>11.7, :Newspaper=>36.8, :Sales=>9.7}, {:""=>51, :TV=>199.8, :Radio=>3.1, :Newspaper=>34.6, :Sales=>11.4}, {:""=>52, :TV

In [58]:
plot = Nyaplot::Plot.new
sc = plot.add_with_df(advertising_data.to_nyaplotdf, :scatter, 'Radio', 'Sales')
plot.x_label("Radio")
plot.y_label("Sales")
plot.show()

#<Nyaplot::Frame:0x007fe2c9be9548 @properties={:panes=>[#<Nyaplot::Plot:0x007fe2c9e52e10 @properties={:diagrams=>[#<Nyaplot::Diagram:0x007fe2c9be9d90 @properties={:type=>:scatter, :options=>{:x=>"Radio", :y=>"Sales"}, :data=>"b0f43243-3cac-43b2-be4c-c4793044158d"}, @xrange=[0, 49.6], @yrange=[1.6, 27]>], :options=>{:x_label=>"Radio", :y_label=>"Sales", :zoom=>true, :width=>700, :xrange=>[0, 49.6], :yrange=>[1.6, 27]}}>], :data=>{"b0f43243-3cac-43b2-be4c-c4793044158d"=>#<Nyaplot::DataFrame:0x007fe2c9c199a0 @name="b0f43243-3cac-43b2-be4c-c4793044158d", @rows=[{:""=>1, :TV=>230.1, :Radio=>37.8, :Newspaper=>69.2, :Sales=>22.1}, {:""=>2, :TV=>44.5, :Radio=>39.3, :Newspaper=>45.1, :Sales=>10.4}, {:""=>3, :TV=>17.2, :Radio=>45.9, :Newspaper=>69.3, :Sales=>9.3}, {:""=>4, :TV=>151.5, :Radio=>41.3, :Newspaper=>58.5, :Sales=>18.5}, {:""=>5, :TV=>180.8, :Radio=>10.8, :Newspaper=>58.4, :Sales=>12.9}, {:""=>6, :TV=>8.7, :Radio=>48.9, :Newspaper=>75, :Sales=>7.2}, {:""=>7, :TV=>57.5, :Radio=>32.8, :Newspaper=>23.5, :Sales=>11.8}, {:""=>8, :TV=>120.2, :Radio=>19.6, :Newspaper=>11.6, :Sales=>13.2}, {:""=>9, :TV=>8.6, :Radio=>2.1, :Newspaper=>1, :Sales=>4.8}, {:""=>10, :TV=>199.8, :Radio=>2.6, :Newspaper=>21.2, :Sales=>10.6}, {:""=>11, :TV=>66.1, :Radio=>5.8, :Newspaper=>24.2, :Sales=>8.6}, {:""=>12, :TV=>214.7, :Radio=>24, :Newspaper=>4, :Sales=>17.4}, {:""=>13, :TV=>23.8, :Radio=>35.1, :Newspaper=>65.9, :Sales=>9.2}, {:""=>14, :TV=>97.5, :Radio=>7.6, :Newspaper=>7.2, :Sales=>9.7}, {:""=>15, :TV=>204.1, :Radio=>32.9, :Newspaper=>46, :Sales=>19}, {:""=>16, :TV=>195.4, :Radio=>47.7, :Newspaper=>52.9, :Sales=>22.4}, {:""=>17, :TV=>67.8, :Radio=>36.6, :Newspaper=>114, :Sales=>12.5}, {:""=>18, :TV=>281.4, :Radio=>39.6, :Newspaper=>55.8, :Sales=>24.4}, {:""=>19, :TV=>69.2, :Radio=>20.5, :Newspaper=>18.3, :Sales=>11.3}, {:""=>20, :TV=>147.3, :Radio=>23.9, :Newspaper=>19.1, :Sales=>14.6}, {:""=>21, :TV=>218.4, :Radio=>27.7, :Newspaper=>53.4, :Sales=>18}, {:""=>22, :TV=>237.4, :Radio=>5.1, :Newspaper=>23.5, :Sales=>12.5}, {:""=>23, :TV=>13.2, :Radio=>15.9, :Newspaper=>49.6, :Sales=>5.6}, {:""=>24, :TV=>228.3, :Radio=>16.9, :Newspaper=>26.2, :Sales=>15.5}, {:""=>25, :TV=>62.3, :Radio=>12.6, :Newspaper=>18.3, :Sales=>9.7}, {:""=>26, :TV=>262.9, :Radio=>3.5, :Newspaper=>19.5, :Sales=>12}, {:""=>27, :TV=>142.9, :Radio=>29.3, :Newspaper=>12.6, :Sales=>15}, {:""=>28, :TV=>240.1, :Radio=>16.7, :Newspaper=>22.9, :Sales=>15.9}, {:""=>29, :TV=>248.8, :Radio=>27.1, :Newspaper=>22.9, :Sales=>18.9}, {:""=>30, :TV=>70.6, :Radio=>16, :Newspaper=>40.8, :Sales=>10.5}, {:""=>31, :TV=>292.9, :Radio=>28.3, :Newspaper=>43.2, :Sales=>21.4}, {:""=>32, :TV=>112.9, :Radio=>17.4, :Newspaper=>38.6, :Sales=>11.9}, {:""=>33, :TV=>97.2, :Radio=>1.5, :Newspaper=>30, :Sales=>9.6}, {:""=>34, :TV=>265.6, :Radio=>20, :Newspaper=>0.3, :Sales=>17.4}, {:""=>35, :TV=>95.7, :Radio=>1.4, :Newspaper=>7.4, :Sales=>9.5}, {:""=>36, :TV=>290.7, :Radio=>4.1, :Newspaper=>8.5, :Sales=>12.8}, {:""=>37, :TV=>266.9, :Radio=>43.8, :Newspaper=>5, :Sales=>25.4}, {:""=>38, :TV=>74.7, :Radio=>49.4, :Newspaper=>45.7, :Sales=>14.7}, {:""=>39, :TV=>43.1, :Radio=>26.7, :Newspaper=>35.1, :Sales=>10.1}, {:""=>40, :TV=>228, :Radio=>37.7, :Newspaper=>32, :Sales=>21.5}, {:""=>41, :TV=>202.5, :Radio=>22.3, :Newspaper=>31.6, :Sales=>16.6}, {:""=>42, :TV=>177, :Radio=>33.4, :Newspaper=>38.7, :Sales=>17.1}, {:""=>43, :TV=>293.6, :Radio=>27.7, :Newspaper=>1.8, :Sales=>20.7}, {:""=>44, :TV=>206.9, :Radio=>8.4, :Newspaper=>26.4, :Sales=>12.9}, {:""=>45, :TV=>25.1, :Radio=>25.7, :Newspaper=>43.3, :Sales=>8.5}, {:""=>46, :TV=>175.1, :Radio=>22.5, :Newspaper=>31.5, :Sales=>14.9}, {:""=>47, :TV=>89.7, :Radio=>9.9, :Newspaper=>35.7, :Sales=>10.6}, {:""=>48, :TV=>239.9, :Radio=>41.5, :Newspaper=>18.5, :Sales=>23.2}, {:""=>49, :TV=>227.2, :Radio=>15.8, :Newspaper=>49.9, :Sales=>14.8}, {:""=>50, :TV=>66.9, :Radio=>11.7, :Newspaper=>36.8, :Sales=>9.7}, {:""=>51, :TV=>199.8, :Radio=>3.1, :Newspaper=>34.6, :Sales=>11.4}, {:""=>52, :TV

In [59]:
plot = Nyaplot::Plot.new
sc = plot.add_with_df(advertising_data.to_nyaplotdf, :scatter, 'Newspaper', 'Sales')
plot.x_label("Newspaper")
plot.y_label("Sales")
plot.show()

#<Nyaplot::Frame:0x007fe2c9fd8d70 @properties={:panes=>[#<Nyaplot::Plot:0x007fe2c9b4b140 @properties={:diagrams=>[#<Nyaplot::Diagram:0x007fe2c9fda3f0 @properties={:type=>:scatter, :options=>{:x=>"Newspaper", :y=>"Sales"}, :data=>"6cd816ef-1c4c-43a2-a22d-0a71d7b4bb85"}, @xrange=[0.3, 114], @yrange=[1.6, 27]>], :options=>{:x_label=>"Newspaper", :y_label=>"Sales", :zoom=>true, :width=>700, :xrange=>[0.3, 114], :yrange=>[1.6, 27]}}>], :data=>{"6cd816ef-1c4c-43a2-a22d-0a71d7b4bb85"=>#<Nyaplot::DataFrame:0x007fe2cb819e20 @name="6cd816ef-1c4c-43a2-a22d-0a71d7b4bb85", @rows=[{:""=>1, :TV=>230.1, :Radio=>37.8, :Newspaper=>69.2, :Sales=>22.1}, {:""=>2, :TV=>44.5, :Radio=>39.3, :Newspaper=>45.1, :Sales=>10.4}, {:""=>3, :TV=>17.2, :Radio=>45.9, :Newspaper=>69.3, :Sales=>9.3}, {:""=>4, :TV=>151.5, :Radio=>41.3, :Newspaper=>58.5, :Sales=>18.5}, {:""=>5, :TV=>180.8, :Radio=>10.8, :Newspaper=>58.4, :Sales=>12.9}, {:""=>6, :TV=>8.7, :Radio=>48.9, :Newspaper=>75, :Sales=>7.2}, {:""=>7, :TV=>57.5, :Radio=>32.8, :Newspaper=>23.5, :Sales=>11.8}, {:""=>8, :TV=>120.2, :Radio=>19.6, :Newspaper=>11.6, :Sales=>13.2}, {:""=>9, :TV=>8.6, :Radio=>2.1, :Newspaper=>1, :Sales=>4.8}, {:""=>10, :TV=>199.8, :Radio=>2.6, :Newspaper=>21.2, :Sales=>10.6}, {:""=>11, :TV=>66.1, :Radio=>5.8, :Newspaper=>24.2, :Sales=>8.6}, {:""=>12, :TV=>214.7, :Radio=>24, :Newspaper=>4, :Sales=>17.4}, {:""=>13, :TV=>23.8, :Radio=>35.1, :Newspaper=>65.9, :Sales=>9.2}, {:""=>14, :TV=>97.5, :Radio=>7.6, :Newspaper=>7.2, :Sales=>9.7}, {:""=>15, :TV=>204.1, :Radio=>32.9, :Newspaper=>46, :Sales=>19}, {:""=>16, :TV=>195.4, :Radio=>47.7, :Newspaper=>52.9, :Sales=>22.4}, {:""=>17, :TV=>67.8, :Radio=>36.6, :Newspaper=>114, :Sales=>12.5}, {:""=>18, :TV=>281.4, :Radio=>39.6, :Newspaper=>55.8, :Sales=>24.4}, {:""=>19, :TV=>69.2, :Radio=>20.5, :Newspaper=>18.3, :Sales=>11.3}, {:""=>20, :TV=>147.3, :Radio=>23.9, :Newspaper=>19.1, :Sales=>14.6}, {:""=>21, :TV=>218.4, :Radio=>27.7, :Newspaper=>53.4, :Sales=>18}, {:""=>22, :TV=>237.4, :Radio=>5.1, :Newspaper=>23.5, :Sales=>12.5}, {:""=>23, :TV=>13.2, :Radio=>15.9, :Newspaper=>49.6, :Sales=>5.6}, {:""=>24, :TV=>228.3, :Radio=>16.9, :Newspaper=>26.2, :Sales=>15.5}, {:""=>25, :TV=>62.3, :Radio=>12.6, :Newspaper=>18.3, :Sales=>9.7}, {:""=>26, :TV=>262.9, :Radio=>3.5, :Newspaper=>19.5, :Sales=>12}, {:""=>27, :TV=>142.9, :Radio=>29.3, :Newspaper=>12.6, :Sales=>15}, {:""=>28, :TV=>240.1, :Radio=>16.7, :Newspaper=>22.9, :Sales=>15.9}, {:""=>29, :TV=>248.8, :Radio=>27.1, :Newspaper=>22.9, :Sales=>18.9}, {:""=>30, :TV=>70.6, :Radio=>16, :Newspaper=>40.8, :Sales=>10.5}, {:""=>31, :TV=>292.9, :Radio=>28.3, :Newspaper=>43.2, :Sales=>21.4}, {:""=>32, :TV=>112.9, :Radio=>17.4, :Newspaper=>38.6, :Sales=>11.9}, {:""=>33, :TV=>97.2, :Radio=>1.5, :Newspaper=>30, :Sales=>9.6}, {:""=>34, :TV=>265.6, :Radio=>20, :Newspaper=>0.3, :Sales=>17.4}, {:""=>35, :TV=>95.7, :Radio=>1.4, :Newspaper=>7.4, :Sales=>9.5}, {:""=>36, :TV=>290.7, :Radio=>4.1, :Newspaper=>8.5, :Sales=>12.8}, {:""=>37, :TV=>266.9, :Radio=>43.8, :Newspaper=>5, :Sales=>25.4}, {:""=>38, :TV=>74.7, :Radio=>49.4, :Newspaper=>45.7, :Sales=>14.7}, {:""=>39, :TV=>43.1, :Radio=>26.7, :Newspaper=>35.1, :Sales=>10.1}, {:""=>40, :TV=>228, :Radio=>37.7, :Newspaper=>32, :Sales=>21.5}, {:""=>41, :TV=>202.5, :Radio=>22.3, :Newspaper=>31.6, :Sales=>16.6}, {:""=>42, :TV=>177, :Radio=>33.4, :Newspaper=>38.7, :Sales=>17.1}, {:""=>43, :TV=>293.6, :Radio=>27.7, :Newspaper=>1.8, :Sales=>20.7}, {:""=>44, :TV=>206.9, :Radio=>8.4, :Newspaper=>26.4, :Sales=>12.9}, {:""=>45, :TV=>25.1, :Radio=>25.7, :Newspaper=>43.3, :Sales=>8.5}, {:""=>46, :TV=>175.1, :Radio=>22.5, :Newspaper=>31.5, :Sales=>14.9}, {:""=>47, :TV=>89.7, :Radio=>9.9, :Newspaper=>35.7, :Sales=>10.6}, {:""=>48, :TV=>239.9, :Radio=>41.5, :Newspaper=>18.5, :Sales=>23.2}, {:""=>49, :TV=>227.2, :Radio=>15.8, :Newspaper=>49.9, :Sales=>14.8}, {:""=>50, :TV=>66.9, :Radio=>11.7, :Newspaper=>36.8, :Sales=>9.7}, {:""=>51, :TV=>199.8, :Radio=>3.1, :Newspaper=>34.6, :Sales=>11.4}, {:"

## Passo 2: Correlação

[[ Volte ao topo ]](#Sumário)

O coeficiente de correlação de Pearson mede o grau da correlação (e a direcção dessa correlação - se positiva ou negativa) entre duas variáveis. Este coeficiente, normalmente representado por $\rho$ assume apenas valores entre -1 e 1.

- $\rho = 1$: Correlação perfeita positiva entre as duas variáveis.
- $\rho = -1$: Correlação negativa perfeita entre as duas variáveis
- $\rho = 0$: As duas variáveis não dependem linearmente uma da outra.

In [60]:
correlation_coefficients = %W(TV Radio Newspaper).each_with_object({}) do |ads, dict|
  dict[ads] = Regression::CorrelationCoefficient.new(advertising_data[ads].to_a, advertising_data['Sales'].to_a)
end

correlation_coefficients.map { |ads, correlation_coefficients| "#{ads} = #{correlation_coefficients.pearson}" }

["TV = 0.7822244248616061", "Radio = 0.5762225745710527", "Newspaper = 0.22829902637616456"]

## Passo 3: Regressão

[[ Volte ao topo ]](#Sumário)

Na regressão linear, coeficientes são estimados usando o **critério dos mínimos quadrados**, que significa encontrar uma linha que minimiza a **soma dos quadrados dos residuos**.

Os elementos presentes na regressão são:
- Valores observados de x e y.
- Modelo dos mínimos quadrados.
- Os resíduos são as distancias entre os valores observados e modelo dos mínimos quadrados.

$$y = \beta_0 + \beta_1x$$

Os coeficientes do modelo estão relacionados com o modelo dos mínimos quadrados de forma:

- $\beta_0$ Constante que representa a interceptação da reta com o eixo vertical
- $\beta_1$ Constante que representa o declive (coeficiente angular) da reta

In [61]:
require 'linear-regression'

linear_regression = Regression::Linear.new(advertising_data['TV'].to_a, advertising_data['Sales'].to_a)
print("y = #{linear_regression.slope}*x + #{linear_regression.intercept}")

y = 0.047536640433019674*x + 7.032593549127708

### Interpretando os coeficientes

Como interpretamos o coeficiente da TV ($\beta_1$)?
- Um aumento de gasto de $1.000 em propagandas televisivas está **associado** com um aumento de 47 vendas.

In [62]:
values = (advertising_data['TV'].min.floor..advertising_data['TV'].max.ceil).to_a

plot = Nyaplot::Plot.new
plot.add_with_df(advertising_data.to_nyaplotdf, :scatter, 'TV', 'Sales')
plot.add(:line, values, values.map { |x| linear_regression.predict(x) })
plot.x_label("TV")
plot.y_label("Sales")
plot.show()

#<Nyaplot::Frame:0x007fe2c89f5d78 @properties={:panes=>[#<Nyaplot::Plot:0x007fe2c8a5c6e0 @properties={:diagrams=>[#<Nyaplot::Diagram:0x007fe2c9d30d20 @properties={:type=>:scatter, :options=>{:x=>"TV", :y=>"Sales"}, :data=>"70ee5050-4ddc-48ce-bdf0-6739b3cc3b82"}, @xrange=[0.7, 296.4], @yrange=[1.6, 27]>, #<Nyaplot::Diagram:0x007fe2c89f6430 @properties={:type=>:line, :options=>{:x=>"data0", :y=>"data1"}, :data=>"11f2c7a1-37b3-42c6-8276-85e1f1c2a895"}, @xrange=[0, 297], @yrange=[7.032593549127708, 21.15097575773455]>], :options=>{:x_label=>"TV", :y_label=>"Sales", :zoom=>true, :width=>700, :xrange=>[0, 297], :yrange=>[1.6, 27]}}>], :data=>{"70ee5050-4ddc-48ce-bdf0-6739b3cc3b82"=>#<Nyaplot::DataFrame:0x007fe2c9d60b10 @name="70ee5050-4ddc-48ce-bdf0-6739b3cc3b82", @rows=[{:""=>1, :TV=>230.1, :Radio=>37.8, :Newspaper=>69.2, :Sales=>22.1}, {:""=>2, :TV=>44.5, :Radio=>39.3, :Newspaper=>45.1, :Sales=>10.4}, {:""=>3, :TV=>17.2, :Radio=>45.9, :Newspaper=>69.3, :Sales=>9.3}, {:""=>4, :TV=>151.5, :Radio=>41.3, :Newspaper=>58.5, :Sales=>18.5}, {:""=>5, :TV=>180.8, :Radio=>10.8, :Newspaper=>58.4, :Sales=>12.9}, {:""=>6, :TV=>8.7, :Radio=>48.9, :Newspaper=>75, :Sales=>7.2}, {:""=>7, :TV=>57.5, :Radio=>32.8, :Newspaper=>23.5, :Sales=>11.8}, {:""=>8, :TV=>120.2, :Radio=>19.6, :Newspaper=>11.6, :Sales=>13.2}, {:""=>9, :TV=>8.6, :Radio=>2.1, :Newspaper=>1, :Sales=>4.8}, {:""=>10, :TV=>199.8, :Radio=>2.6, :Newspaper=>21.2, :Sales=>10.6}, {:""=>11, :TV=>66.1, :Radio=>5.8, :Newspaper=>24.2, :Sales=>8.6}, {:""=>12, :TV=>214.7, :Radio=>24, :Newspaper=>4, :Sales=>17.4}, {:""=>13, :TV=>23.8, :Radio=>35.1, :Newspaper=>65.9, :Sales=>9.2}, {:""=>14, :TV=>97.5, :Radio=>7.6, :Newspaper=>7.2, :Sales=>9.7}, {:""=>15, :TV=>204.1, :Radio=>32.9, :Newspaper=>46, :Sales=>19}, {:""=>16, :TV=>195.4, :Radio=>47.7, :Newspaper=>52.9, :Sales=>22.4}, {:""=>17, :TV=>67.8, :Radio=>36.6, :Newspaper=>114, :Sales=>12.5}, {:""=>18, :TV=>281.4, :Radio=>39.6, :Newspaper=>55.8, :Sales=>24.4}, {:""=>19, :TV=>69.2, :Radio=>20.5, :Newspaper=>18.3, :Sales=>11.3}, {:""=>20, :TV=>147.3, :Radio=>23.9, :Newspaper=>19.1, :Sales=>14.6}, {:""=>21, :TV=>218.4, :Radio=>27.7, :Newspaper=>53.4, :Sales=>18}, {:""=>22, :TV=>237.4, :Radio=>5.1, :Newspaper=>23.5, :Sales=>12.5}, {:""=>23, :TV=>13.2, :Radio=>15.9, :Newspaper=>49.6, :Sales=>5.6}, {:""=>24, :TV=>228.3, :Radio=>16.9, :Newspaper=>26.2, :Sales=>15.5}, {:""=>25, :TV=>62.3, :Radio=>12.6, :Newspaper=>18.3, :Sales=>9.7}, {:""=>26, :TV=>262.9, :Radio=>3.5, :Newspaper=>19.5, :Sales=>12}, {:""=>27, :TV=>142.9, :Radio=>29.3, :Newspaper=>12.6, :Sales=>15}, {:""=>28, :TV=>240.1, :Radio=>16.7, :Newspaper=>22.9, :Sales=>15.9}, {:""=>29, :TV=>248.8, :Radio=>27.1, :Newspaper=>22.9, :Sales=>18.9}, {:""=>30, :TV=>70.6, :Radio=>16, :Newspaper=>40.8, :Sales=>10.5}, {:""=>31, :TV=>292.9, :Radio=>28.3, :Newspaper=>43.2, :Sales=>21.4}, {:""=>32, :TV=>112.9, :Radio=>17.4, :Newspaper=>38.6, :Sales=>11.9}, {:""=>33, :TV=>97.2, :Radio=>1.5, :Newspaper=>30, :Sales=>9.6}, {:""=>34, :TV=>265.6, :Radio=>20, :Newspaper=>0.3, :Sales=>17.4}, {:""=>35, :TV=>95.7, :Radio=>1.4, :Newspaper=>7.4, :Sales=>9.5}, {:""=>36, :TV=>290.7, :Radio=>4.1, :Newspaper=>8.5, :Sales=>12.8}, {:""=>37, :TV=>266.9, :Radio=>43.8, :Newspaper=>5, :Sales=>25.4}, {:""=>38, :TV=>74.7, :Radio=>49.4, :Newspaper=>45.7, :Sales=>14.7}, {:""=>39, :TV=>43.1, :Radio=>26.7, :Newspaper=>35.1, :Sales=>10.1}, {:""=>40, :TV=>228, :Radio=>37.7, :Newspaper=>32, :Sales=>21.5}, {:""=>41, :TV=>202.5, :Radio=>22.3, :Newspaper=>31.6, :Sales=>16.6}, {:""=>42, :TV=>177, :Radio=>33.4, :Newspaper=>38.7, :Sales=>17.1}, {:""=>43, :TV=>293.6, :Radio=>27.7, :Newspaper=>1.8, :Sales=>20.7}, {:""=>44, :TV=>206.9, :Radio=>8.4, :Newspaper=>26.4, :Sales=>12.9}, {:""=>45, :TV=>25.1, :Radio=>25.7, :Newspaper=>43.3, :Sales=>8.5}, {:""=>46, :TV=>175.1, :Radio=>22.5, :Newspaper=>31.5, :Sales=>14.9}, {:""=>47, :TV=>89.7, :Radio=>9.9, :Newspaper=>35.7, :Sales=>10.6}, {:""=>48, :TV=>239.9, :Radio=>41.5, :Newspaper=>18.5, :Sales=>23.2}, {:""=

## Passo 4: Previsão

[[ Volte ao topo ]](#Sumário)
  
Digamos que houve um gasto de $50.000 em propaganda televisiva. Como prever as vendas no mercado?

In [63]:
7.032593549127708 + 0.047536640433019674 * 50

9.40942557077869

Portanto, nós podemos prever uma venda de **9.409 unidades**.